From f30071fbb099f9f6cbec840140630152dab6a6aa Mon Sep 17 00:00:00 2001
From: Andrew Cooper <andrew.cooper3@citrix.com>
Date: Mon, 25 Nov 2024 12:10:16 +0100
Subject: tools/libxs: Track whether we're using a socket or file

It will determine whether to use writev() or sendmsg().

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Juergen Gross <jgross@suse.com>
Reviewed-by: Jason Andryuk <jason.andryuk@amd.com>
master commit: 046efe529e82b8b999d8453d4ea49cb817c3f9b5
master date: 2024-07-23 15:11:27 +0100

diff --git a/tools/libs/store/xs.c b/tools/libs/store/xs.c
index f5e92376ba0d..f16289927e61 100644
--- a/tools/libs/store/xs.c
+++ b/tools/libs/store/xs.c
@@ -65,6 +65,9 @@ struct xs_stored_msg {
 struct xs_handle {
 	/* Communications channel to xenstore daemon. */
 	int fd;
+
+	bool is_socket; /* is @fd a file or socket? */
+
 	Xentoolcore__Active_Handle tc_ah; /* for restrict */
 
 	/*
@@ -140,6 +143,7 @@ static void *read_thread(void *arg);
 
 struct xs_handle {
 	int fd;
+	bool is_socket; /* is @fd a file or socket? */
 	Xentoolcore__Active_Handle tc_ah; /* for restrict */
 	struct list_head reply_list;
 	struct list_head watch_list;
@@ -300,7 +304,9 @@ static struct xs_handle *get_handle(const char *connect_to)
 	if (stat(connect_to, &buf) != 0)
 		goto err;
 
-	if (S_ISSOCK(buf.st_mode))
+	h->is_socket = S_ISSOCK(buf.st_mode);
+
+	if (h->is_socket)
 		h->fd = get_socket(connect_to);
 	else
 		h->fd = get_dev(connect_to);
